File: src/main/kotlin/de/aufmass/kern/Geometrie.kt
package de.aufmass.kern

import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

/**
 * Ein Punkt im **Raumkoordinatensystem**: Ursprung links oben, x nach rechts,
 * y nach hinten, Einheit Meter.
 *
 * y zeigt nach unten – wie in jedem Grundriss und in SVG. Das ist gewöhnungs-
 * bedürftig, wenn man aus der Mathematik kommt, macht aber alles einfacher:
 * Der Bildschirm, der ausgedruckte Plan und die gespeicherten Zahlen meinen
 * dasselbe, ohne dass irgendwo gespiegelt werden muss. Eine Drehung „nach
 * rechts" ist deshalb eine Drehung mit **positivem** Winkel.
 */
data class Punkt(val x: Double, val y: Double)

/**
 * Alles, was sich aus der Form eines Raumes ergibt.
 *
 * Alle Funktionen sind rein: gleicher Raum → gleiches Ergebnis, keine
 * Seiteneffekte, kein Zustand. Damit lassen sie sich in einem Web-Backend,
 * in einem Batch-Export oder in einem Test gleichermaßen verwenden.
 */
object Geometrie {

    // ------------------------------------------------------------ Umriss

    /**
     * Liest den frei erfassten Umriss aus [Raum.umriss].
     *
     * Erwartet wird `[[x,y],[x,y],…]`. Der Parser ist mit Absicht winzig und
     * ohne JSON-Bibliothek geschrieben: Das Paket soll sich ohne Abhängigkeiten
     * übernehmen lassen, und das Format ist eine reine Zahlenliste.
     *
     * **Nachsichtig**: Bei Unsinn kommt `null` zurück statt einer Ausnahme.
     * Ein Tippfehler in einem Feld darf keinen Raum unsichtbar machen – dann
     * gilt eben wieder die Kurzbeschreibung (Rechteck/Trapez).
     */
    fun umrissLesen(text: String): List<Punkt>? {
        if (text.isBlank()) return null
        val zahlen = Regex("-?\\d+(?:\\.\\d+)?").findAll(text)
            .mapNotNull { it.value.toDoubleOrNull() }.toList()
        if (zahlen.size < 6 || zahlen.size % 2 != 0) return null
        val punkte = zahlen.chunked(2).map { Punkt(it[0], it[1]) }
        return if (punkte.size >= 3) punkte else null
    }

    /** Schreibt Eckpunkte in das Format, das [umrissLesen] versteht. */
    fun umrissSchreiben(punkte: List<Punkt>): String =
        punkte.joinToString(",", "[", "]") { "[${runde(it.x)},${runde(it.y)}]" }

    private fun runde(v: Double) = Math.round(v * 100.0) / 100.0

    // ------------------------------------------------------------- Ecken

    /**
     * Von wo bis wo eine Wand der Länge [len] innerhalb der Gesamtbreite [w]
     * verläuft. Die schräge Seite bestimmt, wohin die kürzere Wand rückt: Ist
     * die *linke* Wand schräg, steht die rechte senkrecht – dann enden beide
     * Wände rechts bündig.
     */
    private fun spanne(len: Double, w: Double, s: Schraege): Pair<Double, Double> = when (s) {
        Schraege.LINKS -> (w - len) to w
        Schraege.BEIDSEITIG -> ((w - len) / 2) to ((w + len) / 2)
        else -> 0.0 to len            // RECHTS und KEINE: links bündig
    }

    /**
     * Eckpunkte des Raumes im Uhrzeigersinn.
     *
     * Ist ein Umriss erfasst, gilt er. Sonst entsteht aus Breite hinten,
     * Breite vorne und Tiefe ein Viereck: vorne links, vorne rechts, hinten
     * rechts, hinten links.
     */
    fun ecken(raum: Raum): List<Punkt> {
        umrissLesen(raum.umriss)?.let { return it }
        val hinten = raum.breiteM.toDouble()
        val vorne = raum.breiteVorneM?.toDouble() ?: hinten
        val tiefe = raum.tiefeM.toDouble()
        val w = max(hinten, vorne)
        val (vx0, vx1) = spanne(vorne, w, raum.schraege)
        val (hx0, hx1) = spanne(hinten, w, raum.schraege)
        return listOf(Punkt(vx0, 0.0), Punkt(vx1, 0.0), Punkt(hx1, tiefe), Punkt(hx0, tiefe))
    }

    /** Breite des umschließenden Rechtecks. */
    fun breite(raum: Raum): Double =
        umrissLesen(raum.umriss)?.maxOf { it.x }
            ?: max(raum.breiteM.toDouble(), raum.breiteVorneM?.toDouble() ?: raum.breiteM.toDouble())

    /** Tiefe des umschließenden Rechtecks. */
    fun tiefe(raum: Raum): Double =
        umrissLesen(raum.umriss)?.maxOf { it.y } ?: raum.tiefeM.toDouble()

    /**
     * Grundfläche in m².
     *
     * Für den freien Umriss die **Gaußsche Trapezformel** (Shoelace): Sie gilt
     * für jedes einfache Vieleck, also auch für Nischen und L-förmige Räume.
     * Breite × Tiefe wäre dort zu viel – und wer danach Bodenbelag bestellt,
     * kauft zu viel.
     *
     * Für die Kurzbeschreibung reicht die Trapezformel: Mittellinie × Tiefe.
     */
    fun flaecheM2(raum: Raum): Double {
        umrissLesen(raum.umriss)?.let { p ->
            var summe = 0.0
            for (i in p.indices) {
                val q = p[(i + 1) % p.size]
                summe += p[i].x * q.y - q.x * p[i].y
            }
            return abs(summe) / 2
        }
        val hinten = raum.breiteM.toDouble()
        val vorne = raum.breiteVorneM?.toDouble() ?: hinten
        return (hinten + vorne) / 2 * raum.tiefeM.toDouble()
    }

    /** Umfang in m – Summe aller Wandlängen. */
    fun umfangM(raum: Raum): Double = wandLaengen(raum).sum()

    /** Länge jeder einzelnen Wand, in der Reihenfolge der Ecken. */
    fun wandLaengen(raum: Raum): List<Double> {
        val p = ecken(raum)
        return p.indices.map { i ->
            val q = p[(i + 1) % p.size]
            hypot(q.x - p[i].x, q.y - p[i].y)
        }
    }

    /**
     * Name einer Wand für die Oberfläche. Bei einem Viereck sind die
     * Himmelsrichtungen des Raumes gemeint und leichter zu merken als Zahlen;
     * bei mehr Ecken wird schlicht durchgezählt.
     */
    fun wandName(raum: Raum, index: Int): String {
        val n = ecken(raum).size
        if (n == 4) return listOf("vorne", "rechts", "hinten", "links").getOrElse(index) { "Wand ${index + 1}" }
        return "Wand ${index + 1}"
    }

    // -------------------------------------------------------- Wandpunkte

    /** Anfangs- und Endpunkt einer Wand. */
    fun wand(raum: Raum, index: Int): Pair<Punkt, Punkt> {
        val p = ecken(raum)
        val i = ((index % p.size) + p.size) % p.size
        return p[i] to p[(i + 1) % p.size]
    }

    /** Punkt auf einer Wand, [abstand] Meter von ihrem Anfang entfernt. */
    fun punktAufWand(raum: Raum, index: Int, abstand: Double): Punkt {
        val (a, b) = wand(raum, index)
        val len = hypot(b.x - a.x, b.y - a.y)
        if (len < 1e-9) return a
        val t = abstand.coerceIn(0.0, len) / len
        return Punkt(a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t)
    }

    /** Richtung einer Wand in Grad (0 = nach rechts, im Uhrzeigersinn wachsend). */
    fun wandWinkel(raum: Raum, index: Int): Double {
        val (a, b) = wand(raum, index)
        return Math.toDegrees(Math.atan2(b.y - a.y, b.x - a.x))
    }

    /** Ergebnis der Suche nach der nächstgelegenen Wand. */
    data class WandTreffer(val index: Int, val abstandAufWandM: Double, val entfernungM: Double)

    /**
     * Welche Wand liegt einem Punkt am nächsten, und an welcher Stelle?
     *
     * Gebraucht wird das beim Setzen: Wer eine Steckdose grob in die Nähe
     * einer Wand tippt, meint diese Wand. Ein Schalter schwebt nicht im Raum.
     */
    fun naechsteWand(raum: Raum, x: Double, y: Double): WandTreffer {
        val p = ecken(raum)
        var beste = WandTreffer(0, 0.0, Double.MAX_VALUE)
        for (i in p.indices) {
            val a = p[i]
            val b = p[(i + 1) % p.size]
            val dx = b.x - a.x
            val dy = b.y - a.y
            val l2 = dx * dx + dy * dy
            if (l2 < 1e-9) continue
            val t = (((x - a.x) * dx + (y - a.y) * dy) / l2).coerceIn(0.0, 1.0)
            val px = a.x + dx * t
            val py = a.y + dy * t
            val d = hypot(x - px, y - py)
            if (d < beste.entfernungM) beste = WandTreffer(i, t * Math.sqrt(l2), d)
        }
        return beste
    }

    /** Liegt ein Punkt im Raum? Strahlensatz-Verfahren, gilt für jedes Vieleck. */
    fun imRaum(raum: Raum, x: Double, y: Double): Boolean {
        val p = ecken(raum)
        var drin = false
        var j = p.size - 1
        for (i in p.indices) {
            if ((p[i].y > y) != (p[j].y > y) &&
                x < (p[j].x - p[i].x) * (y - p[i].y) / (p[j].y - p[i].y) + p[i].x
            ) drin = !drin
            j = i
        }
        return drin
    }

    // ------------------------------------------------ Wand mit Öffnungen

    /**
     * Ein Stück Wand, das tatsächlich gemauert ist.
     *
     * [vonM]/[bisM] laufen entlang der Wand, [untenM]/[obenM] über dem Fußboden.
     * Aus diesen Rechtecken baut ein Renderer die Wandscheiben – volle Wand
     * neben der Öffnung, Brüstung darunter, Sturz darüber.
     */
    data class WandStueck(
        val vonM: Double, val bisM: Double,
        val untenM: Double, val obenM: Double
    )

    /**
     * Zerlegt eine Wand in die Stücke, die nach Abzug der Öffnungen übrig
     * bleiben.
     *
     * Das ist der Kern dessen, was eine Tür von einem aufgemalten Symbol
     * unterscheidet: Ohne diese Zerlegung steht man im Modell vor einer
     * geschlossenen Fläche.
     *
     * Überlappende Öffnungen werden zusammengefasst; Öffnungen ausserhalb der
     * Wand werden auf sie begrenzt. Beides kommt bei getippten Werten vor und
     * darf nicht zu Wandstücken mit negativer Länge führen.
     */
    fun wandStuecke(
        raum: Raum,
        wandIndex: Int,
        oeffnungen: List<Oeffnung>
    ): List<WandStueck> {
        val laenge = wandLaengen(raum).getOrElse(wandIndex) { 0.0 }
        val hoehe = raum.hoeheM.toDouble()
        if (laenge <= 0.0 || hoehe <= 0.0) return emptyList()

        val loecher = oeffnungen
            .filter { it.raumId == raum.id && it.wandIndex == wandIndex }
            .map {
                val von = it.abstandM.toDouble().coerceIn(0.0, laenge)
                val bis = (it.abstandM.toDouble() + it.breiteM.toDouble()).coerceIn(0.0, laenge)
                Triple(von, bis, it)
            }
            .filter { it.second > it.first }
            .sortedBy { it.first }

        val stuecke = mutableListOf<WandStueck>()
        var t = 0.0
        for ((von, bis, o) in loecher) {
            if (von > t) stuecke += WandStueck(t, von, 0.0, hoehe)
            val bruestung = min(o.bruestungM.toDouble(), hoehe)
            val sturz = min(o.bruestungM.toDouble() + o.hoeheM.toDouble(), hoehe)
            if (bruestung > 0.0) stuecke += WandStueck(von, bis, 0.0, bruestung)
            if (sturz < hoehe) stuecke += WandStueck(von, bis, sturz, hoehe)
            t = max(t, bis)
        }
        if (t < laenge) stuecke += WandStueck(t, laenge, 0.0, hoehe)
        return stuecke
    }

    // ------------------------------------------------------ Lage im Plan

    /**
     * Wo sitzt ein Einbau im Raumkoordinatensystem, und wie ist er gedreht?
     *
     * An der Wand: auf der Wandlinie, [abrueckenM] in Richtung Raummitte
     * versetzt, damit das Symbol neben der Wand steht und nicht in ihr. Der
     * Winkel ist die Wandrichtung – daran erkennt man im Plan, zu welcher
     * Wand ein Gerät gehört.
     */
    data class Lage(val x: Double, val y: Double, val winkelGrad: Double)

    fun lageImRaum(raum: Raum, einbau: Einbau, abrueckenM: Double = 0.04): Lage {
        if (einbau.befestigung == Befestigung.WAND && einbau.wandIndex != null) {
            val p = punktAufWand(raum, einbau.wandIndex, einbau.abstandM?.toDouble() ?: 0.0)
            val (a, b) = wand(raum, einbau.wandIndex)
            val dx = b.x - a.x
            val dy = b.y - a.y
            val len = hypot(dx, dy).takeIf { it > 1e-9 } ?: 1.0
            // Senkrecht zur Wand abrücken, nicht in Richtung Raummitte: Sonst
            // wandert das Gerät auch ein Stück *an der Wand entlang* und sitzt
            // nicht mehr dort, wo es gemessen wurde. Welche der beiden
            // Senkrechten in den Raum zeigt, entscheidet der Schwerpunkt.
            val e = ecken(raum)
            val mx = e.sumOf { it.x } / e.size
            val my = e.sumOf { it.y } / e.size
            var nx = -dy / len
            var ny = dx / len
            if ((mx - p.x) * nx + (my - p.y) * ny < 0) { nx = -nx; ny = -ny }
            return Lage(
                p.x + nx * abrueckenM,
                p.y + ny * abrueckenM,
                wandWinkel(raum, einbau.wandIndex)
            )
        }
        return Lage(
            einbau.relX.toDouble() * breite(raum),
            einbau.relY.toDouble() * tiefe(raum),
            0.0
        )
    }

    /**
     * Höhe über dem Fußboden, mit der ein Einbau zu zeichnen ist – auch dann,
     * wenn keine erfasst wurde.
     */
    fun hoeheUeberBoden(raum: Raum, einbau: Einbau): Double {
        einbau.hoeheM?.let { return it.toDouble() }
        return when (einbau.befestigung) {
            Befestigung.DECKE -> raum.hoeheM.toDouble()
            Befestigung.BODEN -> 0.0
            else -> Normmasse.hoehe(einbau.art)?.toDouble() ?: 1.20
        }
    }
}

File: web/geometrie.js
/*
 * Geometrie im Browser – Spiegel von `Geometrie.kt`.
 *
 * Warum doppelt? Weil der Editor sofort reagieren muss, während noch getippt
 * wird, und weil der Plan auch ohne Netz stimmen soll. Ein Rundruf zum Server
 * für jede Ziffer wäre weder schnell noch offline-tauglich.
 *
 * **Regel bei Änderungen:** Was hier gerechnet wird, wird auch in
 * `Geometrie.kt` gerechnet – und umgekehrt. Die Kotlin-Tests sind die
 * verbindliche Beschreibung; diese Datei folgt ihnen. Wer eine Formel ändert,
 * ändert beide Stellen und passt den Test an.
 *
 * Koordinaten: Ursprung links oben, x nach rechts, y nach hinten, Meter.
 * Ein Umriss ist ein Array von [x, y]-Paaren im Uhrzeigersinn.
 */

/** Liest den frei erfassten Umriss. Nachsichtig: bei Unsinn kommt null. */
export function umrissLesen(text) {
  if (!text) return null;
  try {
    const p = JSON.parse(text)
      .filter(q => q && q.length >= 2)
      .map(q => [Number(q[0]), Number(q[1])]);
    return p.length >= 3 && p.every(q => isFinite(q[0]) && isFinite(q[1])) ? p : null;
  } catch (_) {
    return null;
  }
}

export function umrissSchreiben(punkte) {
  const r = v => Math.round(v * 100) / 100;
  return JSON.stringify(punkte.map(p => [r(p[0]), r(p[1])]));
}

/** Eckpunkte im Uhrzeigersinn – aus dem Umriss oder aus Breite/Tiefe/Schräge. */
export function ecken(raum) {
  const frei = umrissLesen(raum.umriss ?? raum.outline);
  if (frei) return frei;
  const hinten = Number(raum.breiteM ?? raum.widthM) || 0;
  const vorneRoh = raum.breiteVorneM ?? raum.frontWidthM;
  const vorne = (vorneRoh === null || vorneRoh === undefined) ? hinten : Number(vorneRoh);
  const tiefe = Number(raum.tiefeM ?? raum.lengthM) || 0;
  const w = Math.max(hinten, vorne);
  const schraege = raum.schraege ?? raum.skew;
  const spanne = len => schraege === "LINKS" ? [w - len, w]
    : schraege === "BEIDSEITIG" ? [(w - len) / 2, (w + len) / 2]
      : [0, len];
  const [vx0, vx1] = spanne(vorne), [hx0, hx1] = spanne(hinten);
  return [[vx0, 0], [vx1, 0], [hx1, tiefe], [hx0, tiefe]];
}

/** Breite des umschließenden Rechtecks. */
export function breite(raum) {
  const frei = umrissLesen(raum.umriss ?? raum.outline);
  if (frei) return Math.max(...frei.map(p => p[0]));
  const hinten = Number(raum.breiteM ?? raum.widthM) || 0;
  const vorneRoh = raum.breiteVorneM ?? raum.frontWidthM;
  const vorne = (vorneRoh === null || vorneRoh === undefined) ? hinten : Number(vorneRoh);
  return Math.max(hinten, vorne);
}

/** Tiefe des umschließenden Rechtecks. */
export function tiefe(raum) {
  const frei = umrissLesen(raum.umriss ?? raum.outline);
  return frei ? Math.max(...frei.map(p => p[1])) : (Number(raum.tiefeM ?? raum.lengthM) || 0);
}

/**
 * Grundfläche in m².
 * Freier Umriss: Gaußsche Trapezformel. Kurzform: Mittellinie × Tiefe.
 * Breite × Tiefe wäre bei einer Nische zu viel.
 */
export function flaeche(raum) {
  const frei = umrissLesen(raum.umriss ?? raum.outline);
  if (frei) return flaechePunkte(frei);
  const hinten = Number(raum.breiteM ?? raum.widthM) || 0;
  const vorneRoh = raum.breiteVorneM ?? raum.frontWidthM;
  const vorne = (vorneRoh === null || vorneRoh === undefined) ? hinten : Number(vorneRoh);
  return (hinten + vorne) / 2 * (Number(raum.tiefeM ?? raum.lengthM) || 0);
}

export function flaechePunkte(p) {
  let summe = 0;
  for (let i = 0; i < p.length; i++) {
    const q = p[(i + 1) % p.length];
    summe += p[i][0] * q[1] - q[0] * p[i][1];
  }
  return Math.abs(summe) / 2;
}

/** Länge jeder Wand, in der Reihenfolge der Ecken. */
export function wandLaengen(raum) {
  const p = ecken(raum);
  return p.map((a, i) => {
    const b = p[(i + 1) % p.length];
    return Math.hypot(b[0] - a[0], b[1] - a[1]);
  });
}

/** Nächstgelegene Wand zu einem Punkt (Meter im Raum). */
export function naechsteWand(raum, mx, my) {
  const p = ecken(raum);
  let beste = { index: 0, abstand: 0, entfernung: Infinity };
  for (let i = 0; i < p.length; i++) {
    const a = p[i], b = p[(i + 1) % p.length];
    const dx = b[0] - a[0], dy = b[1] - a[1];
    const l2 = dx * dx + dy * dy || 1;
    const t = Math.max(0, Math.min(1, ((mx - a[0]) * dx + (my - a[1]) * dy) / l2));
    const px = a[0] + dx * t, py = a[1] + dy * t;
    const d = Math.hypot(mx - px, my - py);
    if (d < beste.entfernung) beste = { index: i, abstand: t * Math.sqrt(l2), entfernung: d };
  }
  return beste;
}

/** Liegt ein Punkt im Raum? Gilt für jedes einfache Vieleck. */
export function imRaum(raum, x, y) {
  const p = ecken(raum);
  let drin = false;
  for (let i = 0, j = p.length - 1; i < p.length; j = i++) {
    if ((p[i][1] > y) !== (p[j][1] > y) &&
      x < (p[j][0] - p[i][0]) * (y - p[i][1]) / (p[j][1] - p[i][1]) + p[i][0]) drin = !drin;
  }
  return drin;
}

/**
 * Wandstücke nach Abzug der Öffnungen – volle Wand daneben, Brüstung darunter,
 * Sturz darüber. Ohne diese Zerlegung ist eine Tür nur ein aufgemaltes Symbol.
 */
export function wandStuecke(raum, wandIndex, oeffnungen) {
  const laenge = wandLaengen(raum)[wandIndex] || 0;
  const hoehe = Number(raum.hoeheM ?? raum.heightM) || 0;
  if (laenge <= 0 || hoehe <= 0) return [];
  const loecher = (oeffnungen || [])
    .filter(o => (o.wandIndex ?? o.wallIndex) === wandIndex)
    .map(o => {
      const von = Math.max(0, Math.min(laenge, Number(o.abstandM ?? o.offsetM)));
      const bis = Math.max(0, Math.min(laenge,
        Number(o.abstandM ?? o.offsetM) + Number(o.breiteM ?? o.widthM)));
      return { von, bis, o };
    })
    .filter(l => l.bis > l.von)
    .sort((a, b) => a.von - b.von);

  const stuecke = [];
  let t = 0;
  for (const l of loecher) {
    if (l.von > t) stuecke.push({ von: t, bis: l.von, unten: 0, oben: hoehe });
    const bruestung = Math.min(Number(l.o.bruestungM ?? l.o.sillM) || 0, hoehe);
    const sturz = Math.min(bruestung + (Number(l.o.hoeheM ?? l.o.heightM) || 0), hoehe);
    if (bruestung > 0) stuecke.push({ von: l.von, bis: l.bis, unten: 0, oben: bruestung });
    if (sturz < hoehe) stuecke.push({ von: l.von, bis: l.bis, unten: sturz, oben: hoehe });
    t = Math.max(t, l.bis);
  }
  if (t < laenge) stuecke.push({ von: t, bis: laenge, unten: 0, oben: hoehe });
  return stuecke;
}

/**
 * Lage und Drehung eines Einbaus im Raum.
 *
 * An der Wand: auf der Wandlinie, senkrecht dazu um `abruecken` versetzt.
 * Senkrecht, nicht Richtung Raummitte – sonst wandert das Gerät auch an der
 * Wand entlang und sitzt nicht mehr dort, wo es gemessen wurde.
 */
export function lageImRaum(raum, einbau, abruecken = 0.04) {
  const befestigung = einbau.befestigung ?? einbau.mount;
  const index = einbau.wandIndex ?? einbau.wallIndex;
  const e = ecken(raum);
  if (befestigung === "WAND" && index != null && e.length) {
    const a = e[index % e.length], b = e[(index + 1) % e.length];
    const dx = b[0] - a[0], dy = b[1] - a[1];
    const len = Math.hypot(dx, dy) || 1;
    const t = Math.max(0, Math.min(len, Number(einbau.abstandM ?? einbau.offsetM) || 0));
    const px = a[0] + dx / len * t, py = a[1] + dy / len * t;
    const mx = e.reduce((s, p) => s + p[0], 0) / e.length;
    const my = e.reduce((s, p) => s + p[1], 0) / e.length;
    let nx = -dy / len, ny = dx / len;
    if ((mx - px) * nx + (my - py) * ny < 0) { nx = -nx; ny = -ny; }
    return {
      x: px + nx * abruecken, y: py + ny * abruecken,
      winkel: Math.atan2(dy, dx) * 180 / Math.PI
    };
  }
  return {
    x: Number(einbau.relX) * breite(raum),
    y: Number(einbau.relY) * tiefe(raum),
    winkel: 0
  };
}

/* ------------------------------------------------------------ Polygonzug */

/**
 * Wandliste → Umriss. Start (0,0), Blick nach rechts; nach jeder Wand um
 * ihren Winkel drehen (positiv = rechts).
 *
 * `restluecke` ist die Qualitätsaussage des Aufmaßes: Schließt sich der Zug
 * nicht, steckt ein Messfehler drin.
 */
export function zug(waende) {
  let x = 0, y = 0, richtung = 0;
  const roh = [[0, 0]];
  for (const w of waende) {
    const bogen = richtung * Math.PI / 180;
    x += w.laenge * Math.cos(bogen);
    y += w.laenge * Math.sin(bogen);
    roh.push([x, y]);
    richtung += w.drehung;
  }
  const letzte = roh[roh.length - 1];
  const restluecke = Math.hypot(letzte[0] - roh[0][0], letzte[1] - roh[0][1]);
  const geschlossen = restluecke < 0.02;
  return {
    punkte: normieren(geschlossen ? roh.slice(0, -1) : roh),
    restluecke,
    geschlossen,
    schlusswandErgaenzt: !geschlossen
  };
}

/** Umriss → Wandliste, um einen vorhandenen Raum wandweise zu bearbeiten. */
export function waende(punkte) {
  const n = punkte.length;
  const liste = [];
  for (let i = 0; i < n; i++) {
    const a = punkte[i], b = punkte[(i + 1) % n], c = punkte[(i + 2) % n];
    const laenge = Math.hypot(b[0] - a[0], b[1] - a[1]);
    let drehung = (Math.atan2(c[1] - b[1], c[0] - b[0]) - Math.atan2(b[1] - a[1], b[0] - a[0]))
      * 180 / Math.PI;
    while (drehung > 180) drehung -= 360;
    while (drehung <= -180) drehung += 360;
    liste.push({
      laenge: Math.round(laenge * 100) / 100,
      drehung: Math.round(drehung * 10) / 10
    });
  }
  return liste;
}

/** Auf den Nullpunkt schieben und auf Zentimeter runden. */
export function normieren(punkte) {
  if (!punkte.length) return punkte;
  const minX = Math.min(...punkte.map(p => p[0]));
  const minY = Math.min(...punkte.map(p => p[1]));
  return punkte.map(p => [
    Math.round((p[0] - minX) * 100) / 100,
    Math.round((p[1] - minY) * 100) / 100
  ]);
}

/** Hinweise statt Ausnahmen – der Mensch entscheidet, was ein Fehler ist. */
export function pruefen(punkte) {
  const hinweise = [];
  if (punkte.length < 3) return ["Weniger als drei Ecken – daraus wird kein Raum."];
  punkte.forEach((a, i) => {
    const b = punkte[(i + 1) % punkte.length];
    const l = Math.hypot(b[0] - a[0], b[1] - a[1]);
    if (l < 0.05) hinweise.push(`Wand ${i + 1} ist nur ${Math.round(l * 100)} cm lang – Tippfehler?`);
  });
  if (flaechePunkte(punkte) < 0.5)
    hinweise.push("Fläche unter 0,5 m² – der Zug ist vermutlich in sich verdreht.");
  if (schneidetSichSelbst(punkte))
    hinweise.push("Die Wände kreuzen sich – meist ist eine Drehung falsch herum eingetragen.");
  return hinweise;
}

export function schneidetSichSelbst(p) {
  const n = p.length;
  const rich = (a, b, c) => {
    const v = (b[1] - a[1]) * (c[0] - b[0]) - (b[0] - a[0]) * (c[1] - b[1]);
    return Math.abs(v) < 1e-9 ? 0 : (v > 0 ? 1 : 2);
  };
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (j === i || (j + 1) % n === i || (i + 1) % n === j) continue;
      const a = p[i], b = p[(i + 1) % n], c = p[j], d = p[(j + 1) % n];
      if (rich(a, b, c) !== rich(a, b, d) && rich(c, d, a) !== rich(c, d, b)) return true;
    }
  }
  return false;
}
